### PR TITLE
Allow working with all probe params (kfunc, uprobe)

### LINKED ISSUE #2436: get all arguments values of a function as a string in uprobe and kprobe. something like $$parms in systemtap do.
### Is your feature request related to a problem? Please describe.
I want to print all argument values of many traced functions with uprobe and kprobe. 

### Describe the solution you'd like
provide api to get function's all argument values as a string. some like $$parms in systemtap. https://sourceware.org/systemtap/SystemTap_Beginners_Guide/targetvariables.html

### Describe alternative solutions or features you've considered
Or simpler just provide a variable for the number of argument of each traced function. Then I can just write a loop myself to construct the arguments in a string.

### PR DESCRIPTION
**Pull request by viktormalik**
Allows to use `*args` in multiple constructions such as:
- printing (resolves #2436)
```
# bpftrace -e 'kfunc:vfs_read { print(*args); exit(); }'
Attaching 1 probe...
{ .file = 0xffff9fbe928a7e00, .buf = 0x7ffddb859ee0, .count = 16, .pos = 0xffffb203068afee0 }
```
- storing in map
```
# bpftrace -e 'kfunc:vfs_read { @ = *args; exit(); }'
Attaching 1 probe...

@: { .file = 0xffff9fbe928a7e00, .buf = 0x7ffddb859ee0, .count = 16, .pos = 0xffffb203068afe88 }
```
- map keys
```
# bpftrace -e 'kfunc:vfs_read { @[*args] = count(); } i:ms:100 { exit(); }'
Attaching 2 probes...

[...]
@[{.file=0xffff9fbe8164df00,.buf=0x3434008bd400,.count=1024,.pos=0xffffb2030f24ff00}]: 1
@[{.file=0xffff9fbe4496e400,.buf=0x7ffe84009f10,.count=64,.pos=0xffffb20300757e60}]: 2
@[{.file=0xffff9fbf0c666100,.buf=0x7fff1d7d8577,.count=1,.pos=0x0}]: 2

```

The main change is that probe args are now represented using a special record with `struct <probename>_args` type and hence can be treated as any other record type. Construction of the type is different for kfuncs and uprobes:
- kfunc: the `ctx` pointer points to the struct of args which has exactly the same layout as our representation, so the implementation is straightforward.
- uprobes: args are stored in registers, so when `*args` occurs in the code, the args record is constructed on BPF stack and filled from registers. Note: this is done every time, even when only `args->x` is used (and it would be sufficient to read one register). The reason for this is that creating entire `*args` elegantly fits the current implementation and since there are at most 6 args supported, this should be quite a cheap operation.

This required some minor changes, especially to codegen tests, see commit messages for more details.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by viktormalik** [comment]
ab7681e will go away when #2474 is resolved (do not merge before that).

**Comment by fbs** [comment]
looks good!

**Comment by viktormalik** [comment]
Rebased on master, ready for merge.

**Comment by viktormalik** [comment]
@fbs you didn't put an explicit approval - can we merge this?

@danobi, @ajor want to have a look before that?

Thanks!

**Review by ajor — CHANGES_REQUESTED** [review]
First of all I just want to say that I really like the idea behind this change - the examples look great!

Some of the commits in this PR are simple and unobjectionable and could be merged independently - that would make the core change here a lot less intimidating to go through :)

### Language Design
I know that in the current implementation, `args` is treated as a pointer, but does it need to be?

I don't remember, but I'm assuming I made it like that just because it was easiest - just pass the context pointer directly around and call it "args".

Ideally the bpftrace language won't become full of cruft because of lazy decisions made in the past. It seems to me that a pointer doesn't actually make sense, since we don't allow using it without first dereferencing.

What do you think about allowing use of args without dereferencing? e.g.:
```
kfunc:vfs_read { @a = args }
kfunc:vfs_read { @b = args.file }
```

The old syntax of "args->file" could be kept for backwards compatibility (at least for a while).

### Bugs
I came across a couple of bugs when playing around with this new builtin:
#### Error dereferencing args in tracepoint
```
# ./build/src/bpftrace -e 'tracepoint:syscalls:sys_enter_read { @ = *args }'
Attaching 1 probe...                                                                                      
ERROR: failed to create map: '@': Invalid argument                                                        
Creation of the required BPF maps has failed.                                                             
Make sure you have all the required permissions and are not confined (e.g. like                           
snapcraft does). `dmesg` will likely have useful output for further troubleshooting
```

#### Segfault dereferencing args when function is not defined
```
void foo(int a, int b);
int main() { }
```
```
# ./build/src/bpftrace -e 'uprobe:/tmp/a.out:foo { @ = *args }'
Segmentation fault
```

**Comment by fbs** [comment]
> ### Language Design
> 
> I know that in the current implementation, `args` is treated as a pointer, but does it need to be?
> 
> I don't remember, but I'm assuming I made it like that just because it was easiest - just pass the context pointer directly around and call it "args".
> 
> Ideally the bpftrace language won't become full of cruft because of lazy decisions made in the past. It seems to me that a pointer doesn't actually make sense, since we don't allow using it without first dereferencing.
> 
> What do you think about allowing use of args without dereferencing? e.g.:
> 
> ```
> kfunc:vfs_read { @a = args }
> kfunc:vfs_read { @b = args.file }
> ```
> 
> The old syntax of "args->file" could be kept for backwards compatibility (at least for a while).

Maybe we should just drop the `->` syntax completely? Languages like rust and go are fine without it. I think the type system is good enough to know when we have pointers by now. `$x.a` can just be transformed into `(*$x).a` without the user having to remember?

**Comment by viktormalik** [comment]
> My main concern is unintentionally breaking some part of the C language, like happened when we didn't require the "struct" prefix when casting: #682 A minor concern is confusing C programmers, but I'm sure they'll get used to it.

Well, we'd have to support `->` for backwards compatibility so that shouldn't be an issue. Basically `->` and `.` would have exactly the same effect.

What would change is that users wouldn't be able to rely on the type system to warn them if they use `.` to access pointer variables and vice-versa. But I doubt that anyone relies on that in bpftrace.

**Comment by viktormalik** [comment]
Per agreement with @ajor, I'm splitting this into smaller PRs.

The first three commits are now separated into #2556 and #2557.

**Comment by ajor** [comment]
I'm having second thoughts about the move from "->" to "." now. What would we do about dereferencing pointers when not accessing fields?

C++ and Rust are two languages that mainly use the "." syntax to access struct fields, but that's because they mainly use references instead of pointers. Reading a reference gives the value being pointed to rather than the address it holds (it's implicitly dereferenced). We wouldn't want this behaviour for pointers in bpftrace though, since users will often want and expect an address if they're reading a pointer.

Maybe we should keep pointers and references as separate ideas - we can introduce reference types for higher level language support later, but it feels like this change would be creating something half way between a reference and a pointer. Opinions?

I'm also not sure that it will help with this PR for printing args - if args remains as a pointer type, then printing it will still require `*args`, right? But that dereference was the part I was hoping we could get rid of.

**Comment by danobi** [comment]
I didn't take a very close look at the PR, but the runtime test examples look great.


> I'm having second thoughts about the move from "->" to "." now. What would we do about dereferencing pointers when not accessing fields?

I think simplicity is a good north star for the bpftrace language. While I like the idea of a unifying `.` operator to dereference pointers, I think it is too late to deprecate `->`. We have countless examples and scripts out in the wild. While we could retain support but "soft deprecate it", I suspect that'll just increase confusion. At least for me, I get distrustful of docs/examples that use deprecated features.

> C++ and Rust are two languages that mainly use the "." syntax to access struct fields, but that's because they mainly use references instead of pointers. Reading a reference gives the value being pointed to rather than the address it holds (it's implicitly dereferenced). We wouldn't want this behaviour for pointers in bpftrace though, since users will often want and expect an address if they're reading a pointer.
> 
> Maybe we should keep pointers and references as separate ideas - we can introduce reference types for higher level language support later, but it feels like this change would be creating something half way between a reference and a pointer. Opinions?

I think reference types would require some more discussion, perhaps at an office hours. I have lots of thoughts on that :).

> I'm also not sure that it will help with this PR for printing args - if args remains as a pointer type, then printing it will still require `*args`, right? But that dereference was the part I was hoping we could get rid of.

I don't think we can get rid of it easily. For example rust even supports `*ptr` syntax.

**Comment by viktormalik** [comment]
> > I'm having second thoughts about the move from "->" to "." now. What would we do about dereferencing pointers when not accessing fields?
> 
> I think simplicity is a good north star for the bpftrace language. While I like the idea of a unifying `.` operator to dereference pointers, I think it is too late to deprecate `->`. We have countless examples and scripts out in the wild. While we could retain support but "soft deprecate it", I suspect that'll just increase confusion. At least for me, I get distrustful of docs/examples that use deprecated features.

I agree, deprecating `->` doesn't seem like a good idea. I'd still consider allowing `.` where `->` can be used, though. It may be a bit confusing but we'd basically allow the `.` operator to dereference if applied to a pointer, for the sake of language simplicity.

> I think reference types would require some more discussion, perhaps at an office hours. I have lots of thoughts on that :).

Yes, let's discuss on the next office hours :-)

> > I'm also not sure that it will help with this PR for printing args - if args remains as a pointer type, then printing it will still require `*args`, right? But that dereference was the part I was hoping we could get rid of.
> 
> I don't think we can get rid of it easily. For example rust even supports `*ptr` deref syntax.

But do we need to keep `args` to be a pointer? It's a language builtin so I don't see a problem making it a struct (and perhaps allowing `->` on it, for backwards compatibility).

Anyways, allowing `.` to dereference when necessary would resolve the problem even if `args` remains a pointer.

**Comment by viktormalik** [comment]
Rebased onto master which now includes #2578 which started treating `args` as a struct. This PR now adds capabilities to use `args` in variables and maps (for `kfunc` and `uprobe` probes).

**Review by danobi — APPROVED** [review]
nice!

File: src/ast/irbuilderbpf.cpp
Comment (ajor) [inline src/ast/irbuilderbpf.cpp:1620]: This is causing name collisions - see my top-level comment for details

**Review by ajor — CHANGES_REQUESTED** [review]
I have a couple of questions to help me understand this area better:
**Q1**: Will we able to do something similar to this for tracepoint args in the future or is there some fundamental limitation for them?

**Q2**: Why are uprobe args handled differently from kfunc args? (put on the stack)

I also found a few more problems which we'll need to sort out before merging:

### uprobe arg collisions
```
#include <stdio.h>

void __attribute__((noinline)) bar(int a, int b, int c, int d, int e, int f) {                                                                                        
  printf("hi\n");
}
void __attribute__((noinline)) baz(char g, char h, char i) {                                                                                                          
  printf("hi\n");
}

int main() {                                                                                                                                                          
  bar(11,12,13,14,15,16);
  baz(21, 22, 23);
}
```
`gcc test.c -g -O0`

Works fine with just one uprobe:
```
sudo ./src/bpftrace -e 'uprobe:/tmp/a.out:baz { print(args); }'                                                
Attaching 1 probe...                                                                                                                                                  
{ .g = 21, .h = 22, .i = 23 }                                                                                                                                         
^C
```

Wrong results with multiple uprobes:
```
sudo ./src/bpftrace -e 'uprobe:/tmp/a.out:bar { print(args); } uprobe:/tmp/a.out:baz { print(args); }'         
Attaching 2 probes...                                                                                                                                                 
{ .a = 11, .b = 12, .c = 13, .d = 14, .e = 15, .f = 16 }                                                                                                              
{ .g = 21, .h = 0, .i = 0 }                                                                                                                                           
^C 
```

Crash when uprobe order is swapped:
```
sudo ./src/bpftrace -e 'uprobe:/tmp/a.out:baz { print(args); } uprobe:/tmp/a.out:bar { print(args); }' 
bpftrace: /usr/include/llvm/IR/Instructions.h:922: llvm::Type* llvm::checkGEPType(llvm::Type*): Assertion `Ty && "Invalid GetElementPtrInst indices for type!"' failed.
Aborted
```

### Error message
The error message for uprobes with lots of arguments isn't very nice:
```
sudo ./src/bpftrace -e 'uprobe:/tmp/a.out:foo { print(args); }'
ERROR: Failed to compile: array::at: __n (which is 6) >= _Nm (which is 6)
```

Repro:
```
void __attribute__((noinline)) foo(int a, int b, int c, int d, int e, int f, int g) {                                                                                 
}

int main() {                                                                                                                                                          
  foo(1,2,3,4,5,6,7);                                                                                                                                            
}
```
`gcc test.c -g -O0`

File: src/ast/irbuilderbpf.cpp
Comment (viktormalik) [reply src/ast/irbuilderbpf.cpp:1620]: Nice catch, thanks!

File: src/ast/passes/semantic_analyser.cpp
Comment (viktormalik) [reply src/ast/passes/semantic_analyser.cpp:1489]: Indeed. In fact, only now the original comment underneath matches the condition:
```cpp
        else if (expr->type.IsPtrTy() && expr->type.IsCtxAccess())
        // map functions only accepts a pointer to a element in the stack
```
The check is really about pointers and `args` is not a pointer anymore, so I wouldn't state it explicitly. We can wrap this into some helper like `isNonStackPointer` but I don't think that it would help much.

**Comment by viktormalik** [comment]
Thanks for the review! I added fixup commits to simplify another reviewing and will squash them before merge.

> I have a couple of questions to help me understand this area better: 
>
>**Q1**: Will we able to do something similar to this for tracepoint args in the future or is there some fundamental 
limitation for them?

We should be. Tracepoint args are passed in a similar way to kfuncs, but they have some specialties, like extra fields at the beginning of the struct which we will want to omit or `__data_loc` fields. Once this is merged, I'll open an issue to track support for tracepoints.

> 
> **Q2**: Why are uprobe args handled differently from kfunc args? (put on the stack)

That's b/c for kfuncs, the BPF program gets passed a pointer to a struct containing all the arguments (pretty much exactly what we want). OTOH, uprobe args are passed via registers (the passed context is pointer to `pt_regs`) so we need to build that "struct of args" on stack to have the same codegen representation of `args` for both kfuncs and uprobes (and eventually tracepoints).

File: src/ast/irbuilderbpf.cpp
Comment (ajor) [inline src/ast/irbuilderbpf.cpp:1617]:
```suggestion
  auto type_name = args_type.GetName();
  type_name.erase(0, strlen("struct "));
```
Removes the prefix in-place to avoid creating a temporary std::string with substr(), eliminating an unnecessary malloc/memcpy/free

**Review by ajor — APPROVED** [review]
Nice!

File: src/ast/irbuilderbpf.cpp
Comment (viktormalik) [reply src/ast/irbuilderbpf.cpp]: Good idea, fixed.

**Comment by viktormalik** [comment]
Squashed the fixup commits, ready for merge.

**Pull request merged**
